### ISSUE TITLE
Need detailed install guide for Apple Silicon

### ISSUE THREAD

**Issue by avmak03** [SWivid/F5-TTS#626]
### Checks

- [X] This template is only for usage issues encountered.
- [X] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [X] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [X] I confirm that I am using English to submit this report in order to facilitate communication.

### Environment Details

I tried last week to get it but got too many errors. When finally getting it working, it would analyze the reference audio for a while and then output something. but the file was just silence. 

watched this video but a lot changed since this video was published
https://www.youtube.com/watch?v=a4F9TH56uFo

I downloaded the model from hugging face (model_1200000.pt) and created a ckpts folder to place it. But I got lost after that. I know this is very vague but I genuinely can't figure out how to get this working. I didn't take screenshots from last time or anything to show what I did. But I did download it and ran the model. I got up to the interface and didn't get any sound from the output. This time, I don't know what to do at all.

is creating the ckpts folder even necessary? I see that there is a default model that will be used but I'm following the readme as best as I can and it says to create a folder for a ckpt 

Any help would be appreciated. Thanks in advance

### Steps to Reproduce

1) git clone
2) create a ckpts folder, place the model in
3) [?]

### ✔️ Expected Behavior

Output generated audio

### ❌ Actual Behavior

1) Silence the first time
2) Nothing now. I'm stuck

**Comment by avmak03** [comment]
Thank you for replying. Sorry for the seemingly basic questions but I really don't want to mess this up. This is what I get when going to http://127.0.0.1:7860/ (which is already way different from what I got last time)
![Screenshot 2024-12-13 at 5 16 36 PM](https://github.com/user-attachments/assets/5a2ebfd4-217e-4299-8f40-c4f5dffe8609)

**Comment by manylittlestorms** [comment]
> Thank you for replying. Sorry for the seemingly basic questions but I really don't want to mess this up. This is what I get when going to http://127.0.0.1:7860/ (which is already way different from what I got last time) ![Screenshot 2024-12-13 at 5 16 36 PM](https://private-user-images.githubusercontent.com/96456247/395711799-5a2ebfd4-217e-4299-8f40-c4f5dffe8609.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQxMzI3NDksIm5iZiI6MTczNDEzMjQ0OSwicGF0aCI6Ii85NjQ1NjI0Ny8zOTU3MTE3OTktNWEyZWJmZDQtMjE3ZS00Mjk5LThmNDAtYzRmNWRmZmU4NjA5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEzVDIzMjcyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkwODU1ODI2YmYyMWMzZDJmNzg4MDQ5NjIwNTI0N2Y0N2E0YThmMDA4MDNiMWM1OWJhM2FlNDkxY2YxOTdiOGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gO_i5AZpDrnCuUuu8eLRl6Umk6_Fa2RmHLjJQoPB-LI)

Please make sure you followed these instructions.

```
# Create a python 3.10 conda env (you could also use virtualenv)
conda create -n f5-tts python=3.10
conda activate f5-tts

# Install pytorch with your CUDA version, e.g.
pip install torch==2.3.0+cu118 torchaudio==2.3.0+cu118 --extra-index-url https://download.pytorch.org/whl/cu118
```
If you encounter errors when performing this last instruction, do the following. If not, skip to next part where you paste the git clone:

```
conda install pytorch torchvision -c pytorch
pip install torchaudio
```

Navigate the Terminal to the folder you want to install F5-TTS and paste the following.

```
git clone https://github.com/SWivid/F5-TTS.git
cd F5-TTS
# git submodule update --init --recursive  # (optional, if need bigvgan)
pip install -e .
```

Now, start the UI by pasting in Terminal the following and hit Enter
`f5-tts_infer-gradio`

In the case nothing is generated check if you encounter this error message:
`The operator 'aten::unfold_backward' is not currently implemented for the MPS device

If you encounter that error message, simply Control-C in Terminal to abort the program and paste:
export PYTORCH_ENABLE_MPS_FALLBACK=1

Then, start again with:
`f5-tts_infer-gradio`

You should now see these UI.
<img width="1512" alt="Screenshot 2024-12-14 alle 00 34 19" src="https://github.com/user-attachments/assets/522b2c28-4d4d-4da7-bc40-6b17c9c91d4f" />

**Comment by avmak03** [comment]
> Install pytorch with your CUDA version, e.g.
pip install torch==2.3.0+cu118 torchaudio==2.3.0+cu118 --extra-index-url https://download.pytorch.org/whl/cu118

I'm using a M2 pro Mac mini so I can't do this part. Where can I go from here?

Thanks again!

**Comment by avmak03** [comment]
Update:
I realized that I really did only need 

> conda install pytorch torchvision -c pytorch
> pip install torchaudio

but I used pip for everything:

> pip install torch torchvision
> pip install torchaudio

and I got the same screen. But when I follow the rest of the steps, and try to generate audio, it goes through everything and still outputs no audio. I get this error in the terminal:

> /Users/user/Documents/Newest_F5_TTS/.venv/lib/python3.12/site-packages/gradio/processing_utils.py:738: UserWarning: Trying to convert audio automatically from float32 to 16-bit int format.
> warnings.warn(warning.format(data.dtype))

and here's the spectrogram and the output audio it provides:
![Screenshot 2024-12-14 at 12 49 24 AM](https://github.com/user-attachments/assets/bde4be29-4855-4ec4-8e2b-1d405f5ba1fb)

**Comment by manylittlestorms** [comment]
> Update:
> 
> I realized that I really did only need 
> 
> 
> 
> > conda install pytorch torchvision -c pytorch
> 
> > pip install torchaudio
> 
> 
> 
> but I used pip for everything:
> 
> 
> 
> > pip install torch torchvision
> 
> > pip install torchaudio
> 
> 
> 
> and I got the same screen. But when I follow the rest of the steps, and try to generate audio, it goes through everything and still outputs no audio. I get this error in the terminal:
> 
> 
> 
> > /Users/user/Documents/Newest_F5_TTS/.venv/lib/python3.12/site-packages/gradio/processing_utils.py:738: UserWarning: Trying to convert audio automatically from float32 to 16-bit int format.
> 
> > warnings.warn(warning.format(data.dtype))
> 
> 
> 
> and here's the spectrogram and the output audio it provides:
> 
> ![Screenshot 2024-12-14 at 12 49 24 AM](https://github.com/user-attachments/assets/bde4be29-4855-4ec4-8e2b-1d405f5ba1fb)
> 
> 
> 
> 

@SWivid Can you look into this? Following the steps I wrote in my previous comment I was able to correctly generate audio. But he seems not be able to.
I have an M3 MAX, so I can guarantee my instructions work for Apple Silicon.

**Comment by SWivid** [comment]
@manylittlestorms Sure, I think your instructions are just correct and detailed enough~

@avmak03 If you could got the spectrogram shown, the audio will also be synthesized normally. Make sure you have not suspend during generation in the commandline, just wait for some time.
Also the previously mentioned output is not an error, just `warnings.warn(warning.format(data.dtype))` that could be ignored lol.
If still stuck there, you could provide a full screenshot for us to help better.

**Comment by avmak03** [comment]
@SWivid thank you for helping out too.

I'm still not getting any sound, even after leaving it for a half hour. But the spectrogram is fine.
![Screenshot 2024-12-14 at 5 47 20 PM](https://github.com/user-attachments/assets/9b3bc16d-9d7b-4bbd-a70b-ff5fb3d825a6)


here's all of the output from the terminal:

(.venv) admin@admin_Mac-mini F5-TTS % export PYTORCH_ENABLE_MPS_FALLBACK=1
(.venv) admin@admin_Mac-mini F5-TTS % f5-tts_infer-gradio                 
Building prefix dict from the default dictionary ...
Loading model from cache /var/folders/43/d6yq6wbj4f346mhsxwy2d31c0000gn/T/jieba.cache
Loading model cost 0.335 seconds.
Prefix dict has been built successfully.
Word segmentation module jieba initialized.

Download Vocos from huggingface charactr/vocos-mel-24khz

vocab :  /Users/admin/Documents/Newest_F5_TTS/F5-TTS/src/f5_tts/infer/examples/vocab.txt
token :  custom
model :  /Users/admin/.cache/huggingface/hub/models--SWivid--F5-TTS/snapshots/4dcc16f297f2ff98a17b3726b16f5de5a5e45672/F5TTS_Base/model_1200000.safetensors 

/Users/admin/Documents/Newest_F5_TTS/.venv/lib/python3.12/site-packages/gradio/components/chatbot.py:242: UserWarning: You have not specified a value for the `type` parameter. Defaulting to the 'tuples' format for chatbot messages, but this is deprecated and will be removed in a future version of Gradio. Please set type='messages' instead, which uses openai-style dictionaries with 'role' and 'content' keys.
  warnings.warn(
Starting app...
* Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.

To create a public link, set `share=True` in `launch()`.
Device set to use mps
/Users/admin/Documents/Newest_F5_TTS/.venv/lib/python3.12/site-packages/transformers/models/whisper/generation_whisper.py:512: FutureWarning: The input name `inputs` is deprecated. Please make sure to use `input_features` instead.
  warnings.warn(
You have passed task=transcribe, but also have set `forced_decoder_ids` to [[1, None], [2, 50360]] which creates a conflict. `forced_decoder_ids` will be ignored in favor of task=transcribe.
Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.43.0. You should pass an instance of `EncoderDecoderCache` instead, e.g. `past_key_values=EncoderDecoderCache.from_legacy_cache(past_key_values)`.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.

ref_text   [text was transcribed perfectly. this wasn't an issue regardless of the audio file]

vocab :  /Users/admin/.cache/huggingface/hub/models--SWivid--F5-TTS/snapshots/4dcc16f297f2ff98a17b3726b16f5de5a5e45672/F5TTS_Base/vocab.txt
token :  custom
model :  /Users/admin/.cache/huggingface/hub/models--SWivid--F5-TTS/snapshots/4dcc16f297f2ff98a17b3726b16f5de5a5e45672/F5TTS_Base/model_1200000.safetensors 

gen_text 0 Hello. This is a test.


/Users/admin/Documents/Newest_F5_TTS/.venv/lib/python3.12/site-packages/vocos/spectral_ops.py:46: UserWarning: The operator 'aten::unfold_backward' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:13.)
  return torch.istft(spec, self.n_fft, self.hop_length, self.win_length, self.window, center=True)
/Users/admin/Documents/Newest_F5_TTS/.venv/lib/python3.12/site-packages/gradio/processing_utils.py:738: UserWarning: Trying to convert audio automatically from float32 to 16-bit int format.
  warnings.warn(warning.format(data.dtype))

**Comment by SWivid** [comment]
@avmak03 try two possible solution:
1. first confirm whether only demo works (see badges at top of main page readme in this repo)
2. if not works also, try use a longer gen_text, it's very likely that you have a relatively fast ref_audio and a short gen_text (as shown in the spec in your figs)

**Comment by avmak03** [comment]
@SWivid @manylittlestorms 
Thank you both for your help. I couldn't get it working with gradio but going through CLI worked just fine. I still don't fully understand what the issue was but even with trying it countless times and testing different variables, going through the command line was the only thing that worked out.

Thank you both again!

**Comment by SWivid** [comment]
@avmak03 glad there could be way out
likely it's with gradio stuff 😔, cuz cli works

will close here, feel free to open issue if further questions
